Goodput initial implementation (#32)
triton-inference-server/perf_analyzer: genai-perf/docs/goodput_tutorial.md

   contributors may be used to endorse or promote products derived
   from this software without specific prior written permission.
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->

# Tutorials

## Profile GPT2 running on Triton + vLLM <a id="triton-vllm"></a>

### Run GPT2 on Triton Inference Server using vLLM

<details>
<summary>See instructions</summary>

Run Triton Inference Server with vLLM backend container:

```bash
export RELEASE="yy.mm" # e.g. export RELEASE="24.06"

> docker run -it --net=host --gpus=1 --shm-size=2g --ulimit memlock=-1 --ulimit stack=67108864 nvcr.io/nvidia/tritonserver:${RELEASE}-vllm-python-py3

Reviewer: Could we use vLLM or another approach that does not require several minutes to run? See the tutorials we have that use the OpenAI container as an example.

Author: Updated

  --output-tokens-mean 100 \
  --output-tokens-stddev 0 \
  --output-tokens-mean-deterministic \
  --tokenizer hf-internal-testing/llama-tokenizer \
  --concurrency 4 \
  --measurement-interval 800 \
  --profile-export-file my_profile_export.json \
  --url localhost:8001 \
  --goodput ttft:10 itl:2
```

Example output:

```
                                   LLM Metrics                                    
┏━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┓
┃                Statistic ┃    avg ┃    min ┃    max ┃    p99 ┃    p90 ┃    p75 ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━┩
│ Time to first token (ms) │   9.30 │   5.78 │  20.62 │  20.61 │  18.51 │  10.00 │
│ Inter token latency (ms) │   2.22 │   1.74 │   3.01 │   2.96 │   2.78 │   2.28 │
│     Request latency (ms) │ 256.65 │ 216.91 │ 345.47 │ 345.46 │ 303.17 │ 267.83 │
│   Output sequence length │ 112.41 │ 106.00 │ 125.00 │ 123.71 │ 117.70 │ 115.00 │
│    Input sequence length │ 200.00 │ 200.00 │ 200.00 │ 200.00 │ 200.00 │ 200.00 │
└──────────────────────────┴────────┴────────┴────────┴────────┴────────┴────────┘
Output token throughput (per sec): 1751.15

> Request goodput (per sec): 2.83

Author: Updated

PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->

# Tutorials

## Profile GPT2 running on Triton + vLLM <a id="triton-vllm"></a>

### Run GPT2 on Triton Inference Server using vLLM

<details>
<summary>See instructions</summary>

Run Triton Inference Server with vLLM backend container:

```bash
export RELEASE="yy.mm" # e.g. export RELEASE="24.06"

docker run -it --net=host --gpus=1 --shm-size=2g --ulimit memlock=-1 --ulimit stack=67108864 nvcr.io/nvidia/tritonserver:${RELEASE}-vllm-python-py3

# Install Triton CLI (~5 min):
pip install "git+https://github.com/triton-inference-server/triton_cli@0.0.8"

# Download model:
triton import -m gpt2 --backend vllm

# Run server:
triton start
```

</details>

### Run GenAI-Perf

Run GenAI-Perf from Triton Inference Server SDK container:

```bash
export RELEASE="yy.mm" # e.g. export RELEASE="24.06"

docker run -it --net=host --gpus=1 nvcr.io/nvidia/tritonserver:${RELEASE}-py3-sdk

> # Run GenAI-Perf in the container:

Author: Updated

  --output-tokens-mean-deterministic \
  --tokenizer hf-internal-testing/llama-tokenizer \
  --concurrency 4 \
  --measurement-interval 800 \
  --profile-export-file my_profile_export.json \
  --url localhost:8001 \
  --goodput time_to_first_tokens:10 inter_token_latencies:1.6
```

Example output:

```
                                   LLM Metrics                                    
┏━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┓
┃                Statistic ┃    avg ┃    min ┃    max ┃    p99 ┃    p90 ┃    p75 ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━┩
│ Time to first token (ms) │   5.60 │   4.00 │  13.03 │  12.69 │   7.02 │   5.04 │
│ Inter token latency (ms) │   1.63 │   1.38 │   2.28 │   2.23 │   1.82 │   1.63 │
│     Request latency (ms) │ 184.84 │ 173.10 │ 237.81 │ 237.75 │ 228.81 │ 180.72 │
│   Output sequence length │ 111.43 │ 103.00 │ 126.00 │ 124.55 │ 115.10 │ 114.00 │
│    Input sequence length │ 200.00 │ 200.00 │ 200.00 │ 200.00 │ 200.00 │ 200.00 │
└──────────────────────────┴────────┴────────┴────────┴────────┴────────┴────────┘
Output token throughput (per sec): 1205.26

> Request goodput (per sec): 7.21

Reviewer 2: We are outputting the raw value here.
Do we also want to include a percentage? To give the user of a quick sense of the quality of the run?

> <!--

Reviewer: Please name this file consistently with the other tutorial docs. E.g. "goodput.md"

Author: Please see updated tutorial in a separate PR

Reviewer: Perfect, nice job!

 * Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
 * Neither the name of NVIDIA CORPORATION nor the names of its
   contributors may be used to endorse or promote products derived
   from this software without specific prior written permission.
THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->

# Context

> Goodput, defined as the number of completed requests per second that meet the Service Level Objectives (SLOs), provides an enhanced measure of AI serving performance by accounting for both cost efficiency and user satisfaction.

Reviewer: We use a line length max of 80. These lines should be split up.

Reviewer: You should avoid using terms that are not defined if they are not ubiquitous. Specifically, SLOs are really just constraints that the user specifies. If you can define them first, you can then use them in the documentation. However, since this is more technical documentation, the preferred approach is to use simple language (e.g. constraints) to allow users to skim the documentation to find exactly what they need without reading start to end.

> GenAI-Perf is continually evolving to meet the demands of next-generation GenAI workflows. We are excited to introduce our latest feature: goodput support for measuring Service Level Objectives.

Reviewer: Please shorten these to be more technical. While briefly citing potential uses is good, a lot of these words/sentences don't say anything. Focus on making this high-quality technical documentation rather than high-quality marketing documents.

Also, please remember that this tutorial will need to be up for a long time after it is merged. So time-relevant things like about whether we are introducing a feature would go out in marketing materials, not in our technical documentation.

This new feature empowers you to more precisely evaluate the efficiency and effectiveness of your AI models and services. By measuring goodput, GenAI-Perf delivers deeper insights into how well your services adhere to your defined SLOs, enabling you to optimize both performance and cost efficiency while prioritizing the user experience.

# Tutorials

> This is a tutorial on how to use our goodput feature by examples.

Reviewer: ```suggestion
This is a tutorial on how to use benchmark models using goodput.
```

## Examples

- [LLM Examples](#LLM)

- [Embedding Model Examples](#embeddings)

> ## Profile GPT2 Goodput<a id="LLM"></a>

Reviewer: You should aim to be consistent here. Since you use LLM throughout, the title should be Profile LLM Goodput so that it's obvious that this is the "LLM Examples" section.

I'd recommend using the terminology "Chat Completions" instead though to be more precise and clear with which API is being used.

### Run GPT2 on OpenAI Chat Completions API-compatible server

```bash
docker run -it --net=host --rm --gpus=all vllm/vllm-openai:latest --model gpt2 --dtype float16 --max-model-len 1024
```

> ### Run GenAI-Perf with goodput constraints

Reviewer: All titles should be capitalized.

> <!--

Author: The tutorial needs to be further refined, the main purpose here is to show goodput functionality for LLM, Embeddings and Rankings under three conditions: without goodput-constraints, with valid goodput constraints, and with invalid goodput constraints.

PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->

# Context

Goodput, defined as the number of completed requests per second that meet the Service Level Objectives (SLOs), provides an enhanced measure of AI serving performance by accounting for both cost efficiency and user satisfaction.

GenAI-Perf is continually evolving to meet the demands of next-generation GenAI workflows. We are excited to introduce our latest feature: goodput support for measuring Service Level Objectives.

This new feature empowers you to more precisely evaluate the efficiency and effectiveness of your AI models and services. By measuring goodput, GenAI-Perf delivers deeper insights into how well your services adhere to your defined SLOs, enabling you to optimize both performance and cost efficiency while prioritizing the user experience.

# Tutorials

This is a tutorial on how to use our goodput feature by examples.
## Examples

> - [LLM Examples](#LLM)

Reviewer: You don't need examples for every existing API. You can pick one or two and use them as examples to show how goodput works. By providing every API, we create the need to maintain them long-term and change them as the APIs update.

Author: Updated

Author: Updated

- [Embedding Model Examples](#embeddings)

- [Ranking Model Examples](#rankings)

- [VLM Examples](#VLM)

## Profile GPT2 Goodput running on Triton + vLLM <a id="LLM"></a>

### Run GPT2 on Triton Inference Server using vLLM

Run Triton Inference Server with vLLM backend container:

```bash
export RELEASE="yy.mm" # e.g. export RELEASE="24.06"

docker run -it --net=host --gpus=1 --shm-size=2g --ulimit memlock=-1 --ulimit stack=67108864 nvcr.io/nvidia/tritonserver:${RELEASE}-vllm-python-py3

# Install Triton CLI (~5 min):

> pip install "git+https://github.com/triton-inference-server/triton_cli@0.0.8"

Reviewer: Please use the vLLM container so that we don't need these long-running commands. Also, please go back through all of the feedback the team provided on this PR and respond. If you addressed the issue, then respond so that the person who made the comment can close it.

Author: Updated

  --tokenizer hf-internal-testing/llama-tokenizer \
  --concurrency 2 \
  --measurement-interval 800 \
  --profile-export-file my_profile_export.json \
  --url localhost:8001 \
  --goodput time_to_first_token:8 inter_token_latency:2 request_latency:300
```

Example output:

```
                                   LLM Metrics
┏━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┓
┃                Statistic ┃    avg ┃    min ┃    max ┃    p99 ┃    p90 ┃    p75 ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━┩
│ Time to first token (ms) │   5.50 │   3.78 │  19.25 │  19.20 │   5.86 │   4.98 │
│ Inter token latency (ms) │   1.52 │   1.35 │   1.67 │   1.66 │   1.62 │   1.58 │
│     Request latency (ms) │ 176.02 │ 165.97 │ 197.03 │ 196.99 │ 181.80 │ 177.86 │
│   Output sequence length │ 113.34 │ 107.00 │ 127.00 │ 125.14 │ 118.00 │ 116.00 │
│    Input sequence length │ 200.00 │ 200.00 │ 200.00 │ 200.00 │ 200.00 │ 200.00 │
└──────────────────────────┴────────┴────────┴────────┴────────┴────────┴────────┘
Output token throughput (per sec): 1287.08
Request throughput (per sec): 11.36
Request goodput (per sec): 10.65
```

> ### Run GenAI-Perf with invalid goodput constraints

Reviewer: Please remove this. We don't need to show the invalid case.

                                   LLM Metrics
┏━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┳━━━━━━━━┓
┃                Statistic ┃    avg ┃    min ┃    max ┃    p99 ┃    p90 ┃    p75 ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━╇━━━━━━━━┩
│ Time to first token (ms) │   5.19 │   3.88 │  16.79 │  16.77 │   5.94 │   4.96 │
│ Inter token latency (ms) │   1.56 │   1.44 │   1.69 │   1.68 │   1.63 │   1.60 │
│     Request latency (ms) │ 176.93 │ 170.95 │ 197.70 │ 197.68 │ 177.86 │ 176.21 │
│   Output sequence length │ 111.47 │ 103.00 │ 123.00 │ 122.38 │ 118.80 │ 114.00 │
│    Input sequence length │ 200.00 │ 200.00 │ 200.00 │ 200.00 │ 200.00 │ 200.00 │
└──────────────────────────┴────────┴────────┴────────┴────────┴────────┴────────┘
Output token throughput (per sec): 1259.65
Request throughput (per sec): 11.30
Request goodput (per sec): -1.00
```

> ### Run GenAI-Perf with no goodput constraints

Reviewer: We don't need to show this either.

Author: Updated the tutorial, skip replying the rest of this file.

{"text": "In what state did they film Shrek 2?"}
```

## Start an OpenAI Embeddings-Compatible Server

To start an OpenAI embeddings-compatible server, run the following command:
```bash
docker run -it --net=host --rm --gpus=all vllm/vllm-openai:latest --model intfloat/e5-mistral-7b-instruct --dtype float16 --max-model-len 1024
```

## Run GenAI-Perf
To profile embeddings models using GenAI-Perf, use the following command:

### Run GenAI-Perf with no goodput constraints

```bash
genai-perf profile \
    -m intfloat/e5-mistral-7b-instruct \
    --service-kind openai \
    --endpoint-type embeddings \
    --batch-size 2 \
    --input-file embeddings.jsonl \
    --measurement-interval 1000
```
Example output:

```
                           Embeddings Metrics
┏━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━━┳━━━━━━━┳━━━━━━━┳━━━━━━━┓
┃            Statistic ┃   avg ┃   min ┃    max ┃   p99 ┃   p90 ┃   p75 ┃
┡━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━━╇━━━━━━━╇━━━━━━━╇━━━━━━━┩
│ Request latency (ms) │ 23.51 │ 21.84 │ 200.79 │ 23.17 │ 22.70 │ 22.47 │
└──────────────────────┴───────┴───────┴────────┴───────┴───────┴───────┘
Request throughput (per sec): 42.39
```

> ### Run GenAI-Perf with valid goodput constraints

Reviewer: Every case shown should be valid, so "valid" would be redundant here.
